File: backend/routers/analyze.py
"""AI 个股分析接口（SSE 流式）+ 日志/trace 历史。"""
from __future__ import annotations

from fastapi import APIRouter, Query
from sse_starlette.sse import EventSourceResponse

from apex import analyze as ana
from apex import data as data_mod
from apex import journal
from apex import trace as trace_mod

from backend.core.streaming import stream_callback

router = APIRouter(prefix="/analyze", tags=["analyze"])


@router.get("/run")
def run_analysis(
    ts_code: str = Query(...),
    save: bool = Query(True),
):
    """启动 AI 分析，SSE 流式返回 trace 事件 + 最终结果。

    事件：
      ``trace`` — 分析过程每一步（tool_call / tool_result / assistant_text / context ...）
      ``done``  — 最终分析结果 dict
      ``error`` — 失败
    """
    code = data_mod.normalize_ts_code(ts_code)
    return EventSourceResponse(stream_callback(ana.run, code, save=save))


# ── 日志 / trace ──────────────────────────────────────────────────────────────


journal_router = APIRouter(prefix="/journal", tags=["analyze"])


@journal_router.get("")
def list_all_journal():
    """全部标的全部历史分析（按时间倒序）。供 /journal 跨股历史页。

    ``journal.load_entries()`` 读所有 *.jsonl, **含 position_action**（持仓加减仓建议也是历史记录,
    用户要在历史页看到）。量级 ~数百条, 直接全量返回(前端本地搜索)。

    老条目缺 ``name``(中文名), 这里用全量 name map 兜底回填, 供列表展示。
    """
    entries = journal.load_entries()
    name_map = data_mod.get_name_map()
    if name_map:
        for e in entries:
            if not e.get("name"):
                n = name_map.get(e.get("ts_code", ""))
                if n:
                    e["name"] = n
    entries.sort(
        key=lambda e: e.get("analyzed_at") or e.get("date", ""),
        reverse=True,
    )
    return entries


@journal_router.get("/{ts_code}")
def list_journal(ts_code: str):
    """该股票全部历史分析（按时间倒序，含 position_action 持仓建议）。"""
    code = data_mod.normalize_ts_code(ts_code)
    entries = journal.load_entries(code)
    entries.sort(
        key=lambda e: e.get("analyzed_at") or e.get("date", ""),
        reverse=True,
    )
    return entries


@journal_router.get("/{ts_code}/latest")
def latest_journal(ts_code: str):
    """最近一条分析记录（含 position_action）。供"同步AI到持仓"取最近 advice：
    持仓后最新可能是 position_action（new_stop），前端按 source 分流取 new_stop / price_advice。
    无记录返回 null。
    """
    return journal.load_latest(data_mod.normalize_ts_code(ts_code))


trace_router = APIRouter(prefix="/trace", tags=["analyze"])


@trace_router.get("/{ts_code}/{analyzed_at}")
def get_trace(ts_code: str, analyzed_at: str):
    """某次分析的过程记录（events 列表）。无记录返回 null。"""
    return trace_mod.load_trace(data_mod.normalize_ts_code(ts_code), analyzed_at)

File: frontend/src/api/analyze.ts
/**
 * Analyze API
 *
 * ED11 analyze trace 事件(后端 analyze.py:run 通过 on_progress emit):
 *   { type: "context", name, content }             - 上下文注入
 *   { type: "tool_call", name, args, iteration }   - 调工具
 *   { type: "tool_result", name, result, iteration } - 工具返回
 *   { type: "assistant_text", content, iteration }  - AI 文字
 *   { type: "verdict", ... }                         - 最终 verdict(内嵌在 trace 事件中, done 事件也带)
 *   { error: ... }                                   - 失败
 *
 * 后端 SSE: GET /api/analyze/run?ts_code=xxx&save=true
 *   events: trace(以上) + done(verdict dict) + error
 * done.data = 完整 verdict dict
 */

import { api } from "./client";

export function analyzeRunPath(tsCode: string, save: boolean = true): string {
  // 必须带 /api 前缀 -- Vite proxy 只转发 /api 到 FastAPI,
  // 否则请求落到 dev server SPA fallback,返回 304/403/502(useSSE 报 HTTP <status>)。
  return `/api/analyze/run?ts_code=${encodeURIComponent(tsCode)}&save=${save}`;
}

/**
 * 适配 useSSE 的 fetchFn 形状(参数化路径与方法):
 * useSSE.connect({ path, method, body, signal, fetchFn })
 * fetchFn 收到 { path, method, body, signal } 返回 Promise<Response>
 */
export function analyzeFetchFn() {
  return async (params: {
    path: string;
    method: string;
    body: unknown;
    signal: AbortSignal;
  }): Promise<Response> => {
    return fetch(params.path, {
      method: params.method,
      headers: params.body ? { "Content-Type": "application/json" } : undefined,
      body: params.body ? JSON.stringify(params.body) : undefined,
      signal: params.signal,
    });
  };
}

/** GET /api/journal - 全部标的历史分析(跨股, 倒序)。供 /journal 历史页。 */
export async function getAllJournal(): Promise<unknown[]> {
  return api.get<unknown[]>("/journal");
}

/** GET /api/journal/{ts_code} - 历史 journal 列表 */
export async function getJournal(tsCode: string): Promise<unknown[]> {
  return api.get<unknown[]>(`/journal/${encodeURIComponent(tsCode)}`);
}

/**
 * GET /api/journal/{ts_code}/latest - 最近一条 AI 分析记录。无记录返回 null。
 *
 * 返回 verdict entry(形状见 apex/analyze.py:run),其中 price_advice 含
 * entry/stop_loss/target。用于手动持仓同步 AI advice。
 */
export interface LatestJournal {
  ts_code: string;
  analysis_status?: "completed" | "insufficient_evidence";
  analyzed_at?: string;
  source?: string;
  verdict?: string;
  confidence?: number;
  calibrated_confidence?: number;
  evidence?: Array<string | EvidenceItem>;
  unknowns?: string[];
  research_summary?: string;
  price_advice?: {
    entry?: number | null;
    entry_low?: number | null;
    entry_high?: number | null;
    stop_loss?: number | null;
    target?: number | null;
  };
  position_action?: {
    action?: string;
    add_shares?: number | null;
    trim_shares?: number | null;
    trim_pct?: number | null;
    new_stop?: number | null;
    rationale?: string;
    scale_plan?: unknown[];
  };
}

export interface EvidenceItem {
  id: string;
  fact: string;
  inference?: string;
  evidence_type?: string;
  tool_name?: string;
  source_name?: string;
  source_url?: string;
  published_at?: string | null;
  source_tier?: number;
  entity_matched?: boolean;
  freshness_status?: string;
}

export async function getLatestJournal(
  tsCode: string,
): Promise<LatestJournal | null> {
  return api.get<LatestJournal | null>(
    `/journal/${encodeURIComponent(tsCode)}/latest`,
  );
}

/**
 * GET /api/trace/{ts_code}/{analyzed_at} - 某次分析的完整事件流(trace.jsonl)。
 *
 * 回放历史分析过程用。无记录返回 null(早期分析未落 trace / save=False)。
 * events 形状见 <TraceEventList> 注释(对齐真后端 apex/analyze.py:_emit)。
 * 这里用 unknown[] 而非 TraceEvent[] -- 避免 api 层反向依赖 components 层,
 * 组件层自行断言。
 */
export interface TraceRecord {
  ts_code: string;
  analyzed_at: string;
  events: Record<string, unknown>[];
}

export async function getTrace(
  tsCode: string,
  analyzedAt: string,
): Promise<TraceRecord | null> {
  return api.get<TraceRecord | null>(
    `/trace/${encodeURIComponent(tsCode)}/${encodeURIComponent(analyzedAt)}`,
  );
}
